util/automation.py

```diff
@@ -24,6 +39,139 @@
+        except OSError as ex:
+            print "Unable to start monitoring, check your environment:", ex
+            return
+
+        writer = None
+        with open(self._csv_output, 'w') as f:
+            for line in iter(self._process.stdout.readline, ''):
+                d = self.ingest(line)
+
+                if line.startswith('# '):
+                    if len(self._names) == 0:
+                        self._names = d
+                        writer = csv.DictWriter(f, delimiter=str(','), quotechar=str('"'), fieldnames=d)
+                        writer.writeheader()
+                        continue
+                    if len(self._units) == 0:
+                        self._units = d
+                        continue
+                else:
+                    assert len(self._names) == len(self._units)
+                    assert len(d) == len(self._names)
+                    assert len(d) > 1
+                    writer.writerow(self.merge_line(d))
+                    f.flush()
+
+    def ingest(self, line):
+        return map(lambda x: x.replace('-', '0'), filter(lambda x: len(x) > 0, map(lambda x: x.strip(), line.split(' ')[1:])))
+
+    def merge_line(self, line):
+        return dict(zip(self._names, line))
+
+class GPUUsageChart():
+    def __init__(self, source, basename=None):
+        self._rows    = [ 'pwr', 'temp', 'sm', 'mem']
+        self._titles  = {
+            'pwr':  "Power (W)",
+            'temp': "Temperature (°C)",
+            'sm':   "Streaming Multiprocessors (%)",
+            'mem':  "Memory (%)"
+        }
+        self._data     = { }.fromkeys(self._rows)
+        self._csv      = source
+        self._basename = basename or os.environ.get('ds_gpu_usage_charts', 'gpu_usage_%%s_%d.png' % int(time.time()))
+
+        # This should make sure we start from anything clean.
+        plt.close("all")
+
+        try:
+            self.read()
+            for plot in self._rows:
+                self.produce_plot(plot)
+        except IOError as ex:
+            print "Unable to read", ex
+
+    def append_data(self, row):
+        for bucket, value in row.iteritems():
+            if not bucket in self._rows:
+                continue
+
+            if not self._data[bucket]:
+                self._data[bucket] = {}
+
+            gpu = int(row['gpu'])
+            if not self._data[bucket].has_key(gpu):
+                self._data[bucket][gpu]  = [ value ]
+            else:
+                self._data[bucket][gpu] += [ value ]
+
+    def read(self):
+        print "Reading data from", self._csv
+        with open(self._csv, 'r') as f:
+            for r in csv.DictReader(f):
+                self.append_data(r)
+
+    def produce_plot(self, key, with_spline=True):
+        png = self._basename % (key, )
+        print "Producing plot for", key, "as", png
+        fig, axis = plt.subplots()
+        data = self._data[key]
+        if data is None:
```

Comment:
Be more resilient, we can afford missing data but cannot afford that this would kill the whole run